src/main/studygate/ui.cljs

```diff
@@ -16,64 +16,68 @@
+(defsc TextQuestion [this {:keys [db/id question/displayname question/value question/options question/type]}]
+  {:query [:db/id :question/displayname :question/type
+           :question/options :question/value :question/logicalname]
+   :ident (fn [] [type id])}
   (dom/div #js {:className "question"}
-           (dom/input #js {:className "effect-17"
-                           :type "text"
-                           :value     (or value "")
-                           :onChange #(let [val (-> % .-target .-value)]
-                                        (callback val))})
-           (dom/label nil displayname)
-           (dom/span #js {:className "focus-border"})))
-
-(defn option-input [component callback {:keys [db/id question/displayname question/value question/options]}]
+    (dom/input #js {:className "effect-17"
+                    :type      "text"
+                    :value     (or value "")
+                    :onChange  #(let [val (-> % .-target .-value)]
+                                  (prim/transact! this `[(api/set-question-value ~{:id id :value val})]))})
+    (dom/label nil displayname)
+    (dom/span #js {:className "focus-border"})))
+
+(def ui-text-question (prim/factory TextQuestion {:keyfn :db/id}))
+(defsc OptionQuestion [this {:keys [db/id question/displayname question/value question/options question/type]}]
+  {:query [:db/id :question/displayname :question/type
+           :question/options :question/value :question/logicalname]
+   :ident (fn [] [type id])}
   (dom/div nil
-           (dom/label nil displayname)
-           (dom/ul nil
-                   (map (fn [{:keys [opt-label opt-value]}]
-                          (dom/li #js {:key (str id opt-value)
-                                       :onClick #(callback opt-value)}
-                                  opt-label)) options))))
-
-(defsc SurveyQuestion [this
-                       {:keys [db/id question/displayname question/value
-                               question/logicalname question/type question/options] :as props}
-                       {:keys [value-change] :as computed}]
-  {:query              [:db/id :question/displayname :question/type
-                        :question/options :question/value :question/logicalname]
-   :ident              [:survey-question/by-id :db/id]}
-  (dom/li nil
-          (case type
-            :text (string-input this value-change props)
-            :option (option-input this value-change props))))
+    (dom/label nil displayname)
+    (dom/ul nil
+      (map (fn [{:keys [opt-label opt-value]}]
+             (dom/li #js {:key     (str id opt-value)
+                          :onClick (fn [] (prim/transact! this `[(api/set-question-value ~{:id id :value opt-value})]))}
+               opt-label)) options))))
+
+(def ui-option-question (prim/factory OptionQuestion {:keyfn :db/id}))
+
+(defsc SurveyQuestion [this {:keys [db/id question/type] :as props}]
```

Comment:
Survey question as a union...this breaks your current mutation, since it changes how things normalize. Take a look at the resulting normalized state if you do run this version. Each question type gets its own table.

The power here is each question type can ask for exactly what it needs, and can have custom mutations that deal with just that question type. The downside is that you *have* to consider them separate entity types as a result.